dipy/denoise/localpca.py

```diff
@@ -0,0 +1,114 @@
+import numpy as np
+import scipy as sp
+
+
+def localpca(arr, sigma, patch_radius=1):
+    r"""Local PCA Based Denoising of Diffusion Datasets
+
+    Parameters
+    ----------
+    arr : A 4D array which is to be denoised
+    sigma : float or 3D array
+        standard deviation of the noise estimated from the data
+    patch_radius : The radius of the local patch to
+                be taken around each voxel
+
+    Returns
+    -------
+    denoised_arr : 4D array
+        this is the denoised array of the same size as that of
+        arr (input data)
+
+    References
+    ----------
+    .. [Manjon13] Manjon JV, Coupe P, Concha L, Buades A, Collins DL
+                  Diffusion Weighted Image Denoising Using Overcomplete Local
+                  PCA, PLOS 2013
+    """
+
+    if arr.ndim == 4:
+
+        tou = 2.3 * 2.3 * sigma
+
+        if isinstance(sigma, np.ndarray) and sigma.ndim == 3:
+
+            sigma = (np.ones(arr.shape, dtype=np.float64)
+                     * sigma[..., np.newaxis])
+            tou = (np.ones(arr.shape, dtype=np.float64) * tou[..., np.newaxis])
+        else:
+            sigma = np.ones(arr.shape, dtype=np.float64) * sigma
+            tou = np.ones(arr.shape, dtype=np.float64) * tou
+
+        # loop around and find the 3D patch for each direction at each pixel
+
+        # declare arrays for theta and thetax
+        theta = np.zeros(arr.shape, dtype=np.float64)
+        thetax = np.zeros(arr.shape, dtype=np.float64)
+
+        patch_size = 2 * patch_radius + 1
```

Comment:
The default implementation is using a patch size of 4 directly [1], so you might want to change the keyword to allow this directly as a patch size rather than patch center. You will probably need to play with border effects to get this though to fit exactly with any array.

http://journals.plos.org/plosone/article?id=10.1371%2Fjournal.pone.0073021 section Local PCA Denoising (LPCA)